tokenbucket.go

```diff
@@ -88,12 +88,14 @@ func (t *TokenBucket) Take(ctx context.Context, tokens int64) (time.Duration, er
 	now := t.clock.Now().UnixNano()
 	// Refill the bucket.
 	tokensToAdd := (now - state.Last) / int64(t.refillRate)
+	partialTime := (now - state.Last) % int64(t.refillRate)
 	if tokensToAdd > 0 {
-		state.Last = now
-		if tokensToAdd+state.Available <= t.capacity {
+		if tokensToAdd+state.Available < t.capacity {
 			state.Available += tokensToAdd
+			state.Last = now - partialTime
 		} else {
 			state.Available = t.capacity
+			state.Last = now
```

Comment:
If `state.Available` would be >= `t.capacity`, that would indicate that the bucket would have been full in the time between the current request and previous one. The reason this is `state.Last = now` instead of accounting for partial time is that accounting for partial time would add partial tokens to an already full bucket.

1. rate = 100ms
2. Request 1 occurs at t = 0ms, consuming a token.
3. Request 2 occurs at t = 150ms, consuming and adding a token. The bucket would have been full at t = 100ms.
4. Request 3 occurs at t = 230ms, consuming a token. A token should not be added because the bucket didn't start to refill until t = 150ms, which is 80ms prior.

In the above example, if the partial time was taken into account with Request 2, an extra token would have been added with Request 3 when it shouldn't have.

Reply:
Yeah, sounds good. Please update the tests so that this case is captured.